src/main/kotlin/org/jitsi/nlj/rtp/TransportCcEngine.kt

```diff
@@ -100,37 +104,47 @@ class TransportCcEngine(
             localReferenceTime = now
         }
 
-        // We have to remember the oldest known sequence number here, as we
-        // remove from sentPacketDetails inside this loop
-        val oldestKnownSeqNum = synchronized(sentPacketsSyncRoot) { sentPacketDetails.oldestEntry() }
         for (packetReport in tccPacket) {
             val tccSeqNum = packetReport.seqNum
             val packetDetail = synchronized(sentPacketsSyncRoot) {
-                sentPacketDetails.remove(tccSeqNum)
+                sentPacketDetails.get(tccSeqNum)
             }
 
             if (packetDetail == null) {
                 if (packetReport is ReceivedPacketReport) {
+                    currArrivalTimestamp += packetReport.deltaDuration
                     missingPacketDetailSeqNums.add(tccSeqNum)
                 }
                 continue
             }
 
             when (packetReport) {
-                is UnreceivedPacketReport ->
-                    bandwidthEstimator.processPacketLoss(now, packetDetail.packetSendTime, tccSeqNum)
-
+                is UnreceivedPacketReport -> {
+                    if (packetDetail.state == PacketDetailState.unreported) {
+                        bandwidthEstimator.processPacketLoss(now, packetDetail.packetSendTime, tccSeqNum)
+                        packetDetail.state = PacketDetailState.reportedLost
+                    }
+                }
                 is ReceivedPacketReport -> {
                     currArrivalTimestamp += packetReport.deltaDuration
 
-                    val arrivalTimeInLocalClock = currArrivalTimestamp - Duration.between(localReferenceTime, remoteReferenceTime)
-
-                    bandwidthEstimator.processPacketArrival(
-                        now, packetDetail.packetSendTime, arrivalTimeInLocalClock, tccSeqNum, packetDetail.packetLength)
+                    when (packetDetail.state) {
+                        PacketDetailState.unreported, PacketDetailState.reportedLost -> {
+                            val arrivalTimeInLocalClock = currArrivalTimestamp - Duration.between(localReferenceTime, remoteReferenceTime)
+
+                            /* TODO: BandwidthEstimator should have an API for "previously reported lost packet has arrived"
+                             * for the reportedLost case. */
```

Comment:
I wonder if we should just not report unreceived as lost?  The spec does say:

> Packets with status "Packet not received" should not necessarily be
> interpreted as lost

https://tools.ietf.org/html/draft-holmer-rmcat-transport-wide-cc-extensions-01#section-3.1.1

Reply:
A packet sequence gap in a TCC was either not sent, lost, or re-ordered.  If it wasn't sent we won't have a `PacketStatus` for it; if it was re-ordered and shows up in a later TCC, we'll do the `reportedLost` -> `reportedReceived` transition.  So I think this is a fair description of the situation.

Reply:
Ok, fair enough.